{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# The Characterization of Single Amino Acid Polymorphism of UafA   \n",
    "@author Changhua Yu  \n",
    "@date 181009"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Overview \n",
    "- 180523: Literature Review  \n",
    "- 180601: Pymol Visualization of the mutated sites  \n",
    "    - The overall mapping of mutated sites\n",
    "    - The overall mapping of mutated sites with site frequency as spectrum  \n",
    "    - Mutated haplotypes visualization   \n",
    "    - The potential functioning sites according literature  \n",
    "    - The overlapping between mutated sites and potential functioning sites\n",
    "- 180613: \n",
    "    1. Better visualization on superimposition between annotated regions and mutated sites\n",
    "        - Generate the overlapping of all the mutated sites with each of the functional region        \n",
    "        - Finish the labelling of func region in Pymol: As the Pymol display a label for each of the selected residues\n",
    "          atoms, use a color scheme table instead\n",
    "    2. Generate useful statistics from the sqlite3 outputs for the functional predictions   \n",
    "        - Generate tables to answer: which site, AA change, site-AA pair and which region has the most frequent change\n",
    "        - Output a correlation count matrix with a side length of N (N is the number of all distinct AA- site pairs)\n",
    "        - Visualize spatial relationship of the mutated sites\n",
    "    3. Structural prediction on N1 Region by online servers:\n",
    "        - I-TASSER\n",
    "        - Phyre2--homologous\n",
    "        - Robbeta -- direct folded, only good for the shorter sequences ,N1  only e.g \n",
    "        - Pfam -- alignment based prediction\n",
    "        - Research more on the prediction tool relating to single amino acid mutagenesis\n",
    "- 180626:\n",
    "    1. Better visualization on superimposition between annotated regions and mutated sites\n",
    "        - Generate useful statistics from the sqlite3 outputs for the functional predictions   \n",
    "        - Generate tables to answer: which site, AA change, site-AA pair and which region has the most frequent change \n",
    "        - Output a correlation count matrix with a side length of N (N is the number of all distinct AA- site pairs)\n",
    "        Visualize spatial relationship of the mutated sites\n",
    "- 180727:\n",
    "    1. More Visualization\n",
    "        - Generate the superimposition of the each haplotype with each annotated regions\n",
    "        - Output a chart for assistive inspection \n",
    "    2. Re-output the tables\n",
    "        - Fix the omission of deletion \n",
    "        - Fix the calculation of correlated occurence\n",
    "\n",
    "The project is not done and this doc will continue as a documentation and implementation demo of the function from now on"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Fix the calculation of correlated occurence\n",
    "- change FreqCorrMatrix.py\n",
    "- the frequency should be (p(A1,A2))^2/p(A1)p(A2) from the mute_corr_count_2_180727 table\n",
    "- insert into mute_corr_freq\n",
    "- pattern_id| word_length| line| hap_id| corr_freq\n",
    "- compare with the raw_count table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pattern_id</th>\n",
       "      <th>word_length</th>\n",
       "      <th>line</th>\n",
       "      <th>hap_id</th>\n",
       "      <th>corr_freq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5027</td>\n",
       "      <td>2</td>\n",
       "      <td>40,37</td>\n",
       "      <td>45,57,26,59,29,63</td>\n",
       "      <td>0.857143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5047</td>\n",
       "      <td>2</td>\n",
       "      <td>49,37</td>\n",
       "      <td>45,57,26,59,29,63</td>\n",
       "      <td>0.857143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5253</td>\n",
       "      <td>2</td>\n",
       "      <td>90,98</td>\n",
       "      <td>2,3,6,7,8,9,10,11,12,13,14,15,16,19,22,23,24,2...</td>\n",
       "      <td>0.770833</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4763</td>\n",
       "      <td>2</td>\n",
       "      <td>72,65</td>\n",
       "      <td>57,59,45</td>\n",
       "      <td>0.750000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4793</td>\n",
       "      <td>2</td>\n",
       "      <td>74,65</td>\n",
       "      <td>57,59,45</td>\n",
       "      <td>0.750000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5028</td>\n",
       "      <td>2</td>\n",
       "      <td>40,32</td>\n",
       "      <td>45,57,26,59,29,63</td>\n",
       "      <td>0.750000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>5029</td>\n",
       "      <td>2</td>\n",
       "      <td>40,46</td>\n",
       "      <td>45,57,26,59,29,63</td>\n",
       "      <td>0.750000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>5030</td>\n",
       "      <td>2</td>\n",
       "      <td>40,51</td>\n",
       "      <td>45,57,26,59,29,63</td>\n",
       "      <td>0.750000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>5048</td>\n",
       "      <td>2</td>\n",
       "      <td>49,32</td>\n",
       "      <td>45,57,26,59,29,63</td>\n",
       "      <td>0.750000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>5049</td>\n",
       "      <td>2</td>\n",
       "      <td>49,46</td>\n",
       "      <td>45,57,26,59,29,63</td>\n",
       "      <td>0.750000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   pattern_id  word_length   line  \\\n",
       "0        5027            2  40,37   \n",
       "1        5047            2  49,37   \n",
       "2        5253            2  90,98   \n",
       "3        4763            2  72,65   \n",
       "4        4793            2  74,65   \n",
       "5        5028            2  40,32   \n",
       "6        5029            2  40,46   \n",
       "7        5030            2  40,51   \n",
       "8        5048            2  49,32   \n",
       "9        5049            2  49,46   \n",
       "\n",
       "                                              hap_id  corr_freq  \n",
       "0                                  45,57,26,59,29,63   0.857143  \n",
       "1                                  45,57,26,59,29,63   0.857143  \n",
       "2  2,3,6,7,8,9,10,11,12,13,14,15,16,19,22,23,24,2...   0.770833  \n",
       "3                                           57,59,45   0.750000  \n",
       "4                                           57,59,45   0.750000  \n",
       "5                                  45,57,26,59,29,63   0.750000  \n",
       "6                                  45,57,26,59,29,63   0.750000  \n",
       "7                                  45,57,26,59,29,63   0.750000  \n",
       "8                                  45,57,26,59,29,63   0.750000  \n",
       "9                                  45,57,26,59,29,63   0.750000  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import sqlite3\n",
    "import pandas as pd\n",
    "# connect to sqlite3 local database\n",
    "\n",
    "db_file = \"../db/uafa.db\"\n",
    "conn = sqlite3.connect(db_file)\n",
    "cur = conn.cursor()\n",
    "cmd1 = '''select * from mute_corr_freq where corr_freq < 1.0 order by -1*corr_freq  limit 10;'''\n",
    "df1 = pd.read_sql_query(cmd1, conn)\n",
    "df1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pattern_id</th>\n",
       "      <th>word_length</th>\n",
       "      <th>line</th>\n",
       "      <th>hap_id</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5253</td>\n",
       "      <td>2</td>\n",
       "      <td>90,98</td>\n",
       "      <td>2,3,6,7,8,9,10,11,12,13,14,15,16,19,22,23,24,2...</td>\n",
       "      <td>37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5233</td>\n",
       "      <td>2</td>\n",
       "      <td>1,3</td>\n",
       "      <td>35,37,39,40,42,45,46,47,48,51,53,54,55,56,57,5...</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5234</td>\n",
       "      <td>2</td>\n",
       "      <td>1,5</td>\n",
       "      <td>35,37,39,40,42,45,46,47,48,51,53,54,55,56,57,5...</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5235</td>\n",
       "      <td>2</td>\n",
       "      <td>1,100</td>\n",
       "      <td>35,37,39,40,42,45,46,47,48,51,53,54,55,56,57,5...</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5236</td>\n",
       "      <td>2</td>\n",
       "      <td>1,102</td>\n",
       "      <td>35,37,39,40,42,45,46,47,48,51,53,54,55,56,57,5...</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5238</td>\n",
       "      <td>2</td>\n",
       "      <td>1,98</td>\n",
       "      <td>35,37,39,40,42,45,46,47,48,51,53,54,55,56,57,5...</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>5239</td>\n",
       "      <td>2</td>\n",
       "      <td>3,5</td>\n",
       "      <td>35,37,39,40,42,45,46,47,48,51,53,54,55,56,57,5...</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>5240</td>\n",
       "      <td>2</td>\n",
       "      <td>3,100</td>\n",
       "      <td>35,37,39,40,42,45,46,47,48,51,53,54,55,56,57,5...</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>5241</td>\n",
       "      <td>2</td>\n",
       "      <td>3,102</td>\n",
       "      <td>35,37,39,40,42,45,46,47,48,51,53,54,55,56,57,5...</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>5243</td>\n",
       "      <td>2</td>\n",
       "      <td>3,98</td>\n",
       "      <td>35,37,39,40,42,45,46,47,48,51,53,54,55,56,57,5...</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   pattern_id  word_length   line  \\\n",
       "0        5253            2  90,98   \n",
       "1        5233            2    1,3   \n",
       "2        5234            2    1,5   \n",
       "3        5235            2  1,100   \n",
       "4        5236            2  1,102   \n",
       "5        5238            2   1,98   \n",
       "6        5239            2    3,5   \n",
       "7        5240            2  3,100   \n",
       "8        5241            2  3,102   \n",
       "9        5243            2   3,98   \n",
       "\n",
       "                                              hap_id  count  \n",
       "0  2,3,6,7,8,9,10,11,12,13,14,15,16,19,22,23,24,2...     37  \n",
       "1  35,37,39,40,42,45,46,47,48,51,53,54,55,56,57,5...     24  \n",
       "2  35,37,39,40,42,45,46,47,48,51,53,54,55,56,57,5...     24  \n",
       "3  35,37,39,40,42,45,46,47,48,51,53,54,55,56,57,5...     24  \n",
       "4  35,37,39,40,42,45,46,47,48,51,53,54,55,56,57,5...     24  \n",
       "5  35,37,39,40,42,45,46,47,48,51,53,54,55,56,57,5...     24  \n",
       "6  35,37,39,40,42,45,46,47,48,51,53,54,55,56,57,5...     24  \n",
       "7  35,37,39,40,42,45,46,47,48,51,53,54,55,56,57,5...     24  \n",
       "8  35,37,39,40,42,45,46,47,48,51,53,54,55,56,57,5...     24  \n",
       "9  35,37,39,40,42,45,46,47,48,51,53,54,55,56,57,5...     24  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cmd2 = '''select * from mute_corr_count_2_180727 where count > 0 order by -1*count limit 10;'''\n",
    "df2 = pd.read_sql_query(cmd2, conn)\n",
    "df2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Mutual Information Calculation of correlated text\n",
    "- After learning the concept in BioE131, it seems that mutual information is what John-Marc initially suggests  \n",
    "- A measure of how correlated the two objects are  \n",
    "- I(x,y) = log2[P(x,y)/(P(x)*P(y))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>patter_id</th>\n",
       "      <th>word_length</th>\n",
       "      <th>line</th>\n",
       "      <th>hap_id</th>\n",
       "      <th>corr_MI</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>107</td>\n",
       "      <td>2</td>\n",
       "      <td>4,11</td>\n",
       "      <td>28</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>321</td>\n",
       "      <td>2</td>\n",
       "      <td>7,41</td>\n",
       "      <td>40</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>406</td>\n",
       "      <td>2</td>\n",
       "      <td>9,13</td>\n",
       "      <td>25</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1164</td>\n",
       "      <td>2</td>\n",
       "      <td>21,30</td>\n",
       "      <td>63</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1166</td>\n",
       "      <td>2</td>\n",
       "      <td>21,35</td>\n",
       "      <td>63</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1175</td>\n",
       "      <td>2</td>\n",
       "      <td>21,61</td>\n",
       "      <td>63</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1178</td>\n",
       "      <td>2</td>\n",
       "      <td>21,71</td>\n",
       "      <td>63</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1179</td>\n",
       "      <td>2</td>\n",
       "      <td>21,73</td>\n",
       "      <td>63</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1180</td>\n",
       "      <td>2</td>\n",
       "      <td>21,75</td>\n",
       "      <td>63</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1181</td>\n",
       "      <td>2</td>\n",
       "      <td>21,80</td>\n",
       "      <td>63</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   patter_id  word_length   line hap_id  corr_MI\n",
       "0        107            2   4,11     28      1.0\n",
       "1        321            2   7,41     40      1.0\n",
       "2        406            2   9,13     25      1.0\n",
       "3       1164            2  21,30     63      1.0\n",
       "4       1166            2  21,35     63      1.0\n",
       "5       1175            2  21,61     63      1.0\n",
       "6       1178            2  21,71     63      1.0\n",
       "7       1179            2  21,73     63      1.0\n",
       "8       1180            2  21,75     63      1.0\n",
       "9       1181            2  21,80     63      1.0"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import sqlite3\n",
    "import pandas as pd\n",
    "# connect to sqlite3 local database\n",
    "\n",
    "db_file = \"../db/uafa.db\"\n",
    "conn = sqlite3.connect(db_file)\n",
    "cur = conn.cursor()\n",
    "cmd1 = '''select * from mute_corr_MI order by -1 * corr_MI  limit 10;'''\n",
    "df1 = pd.read_sql_query(cmd1, conn)\n",
    "df1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compare raw count, Bayesian and MI result on graph\n",
    "- The in both chart and colored pymol model\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
